### PR TITLE
rocm-opencl-icd: init at 3.5.0

### PR DESCRIPTION
**Pull request by danieldk**
###### Motivation for this change

This PR adds the open source AMD ROCm packages for OpenCL support. All the heavy lifting was done by @acowley in the [nixos-rocm](https://github.com/nixos-rocm/nixos-rocm) project. All credit for this work should go to Anthony. I only cleaned up the derivations a little to conform more to nixpkgs guidelines.

I did some very preliminary testing by trying Blender's OpenCL support. Blender seems to pick it up fine.

@acowley I had some questions:

* Should I add you as maintainer of these packages?
* Is `patchelf` in rocm-opencl-runtime still necessary? `clinfo` seems to work fine without? I guess we could also consider removing the `clinfo` binary, since it is already provided by the `clinfo` package in `nixpkgs`.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/llvm/rocm/default.nix
Comment (acowley) [inline pkgs/development/compilers/llvm/rocm/default.nix:17]: Can we automate the version number part of this path? It is a nuisance to maintain because we get version bumps independent of public releases, requiring that a person make sure this line is in sync with wherever the rocm fork is with respect to upstream's development branch.

**Comment by acowley** [comment]
@danieldk You are selling yourself short: this cleanup you've done is fantastic! So many things that I stumbled around trying to figure out and fix, you've made comprehensible. I would very much like to upstream these changes into `nixos-rocm` if we are indeed going to continue using the overlay for development.

High-level comments:

- Should we rename `rocr` and `roct`? These are the prefixes of their upstream repositories, but we could name them `rocm-runtime` and `rocm-thunk` or something. Confusion with `rocclr` seems hard to avoid, and I like the short names, but if we did want to change them now is a good time.

- Do we need `llvm-outputs.patch`? I understand it is from the `nixpkgs` LLVM packaging, but I'm not sure I understand how it comes into play.

- I think we should include `hip-clang` here, if possible. It is a small build, and arguably the key piece of the ROCm stack since it makes building CUDA source so straightforward. It used to be quite the circus to build it, but as of 3.5 it is a much better story.

- How do you want to handle `rocr-ext`? This is very often a key issue for OpenCL development since it is required for Image support. The huge issue is that it is not open source. If we don't have it here, I worry that would-be nixpkgs users of OpenCL would be forced to turn to `nixos-rocm` before they are able to do anything. Additionally, support for `darktable`, for example, requires that image support. A large swath of the broader OpenCL ecosystem needs image support, and we do have it in `nixos-rocm`, but with the caveat that there is the proprietary software question.

- We probably need some documentation associated with configuring NixOS to use ROCm. Things like having to reboot to get OpenCL working (I think this is still the case), are perennial gotchas. Perhaps this belongs in Chapter 16 of the nixpkgs manual.

Again, I am so impressed by and grateful for the work you've done here in cleaning up these derivations. What was before an instance of shipping the MVP is here something much closer to a designed and thought-through artifact.

File: pkgs/development/compilers/llvm/rocm/default.nix
Comment (danieldk) [reply pkgs/development/compilers/llvm/rocm/default.nix]: Changed this to get the version from `clang` itself.

File: pkgs/development/libraries/rocm-opencl-runtime/default.nix
Comment (danieldk) [reply pkgs/development/libraries/rocm-opencl-runtime/default.nix]: Removed, except for the one comment about fixing up the ICD path.

**Comment by danieldk** [comment]
Thanks for all the feedback, I have pushed updated commits.

> I would very much like to upstream these changes into `nixos-rocm` if we are indeed going to continue using the overlay for development.

Once this is merged in nixpkgs I can do a PR for `nixos-rocm` to incorporate the changes.

>     * Should we rename `rocr` and `roct`? These are the prefixes of their upstream repositories, but we could name them `rocm-runtime` and `rocm-thunk` or something. Confusion with `rocclr` seems hard to avoid, and I like the short names, but if we did want to change them now is a good time.

That's indeed a lot less confusing, so I renamed these two derivations.

>     * Do we need `llvm-outputs.patch`? I understand it is from the `nixpkgs` LLVM packaging, but I'm not sure I understand how it comes into play.

I think we still need this, because dynamic libraries are in the separate `lib` output. It is probably useful for reducing the closure size, since the final OpenCL runtime only uses the `lib` output:

```
❯ nix-store -qR result | grep llvm
/nix/store/cy5c6z1pxwr2r40wslqhc4k732vp7nwd-rocm-llvm-3.5.1-lib
```
>     * I think we should include `hip-clang` here, if possible. It is a small build, and arguably the key piece of the ROCm stack since it makes building CUDA source so straightforward. It used to be quite the circus to build it, but as of 3.5 it is a much better story.

Sounds good. I think this PR is already quite big. So once it's merged, I can look at `hip-clang` to.

>     * How do you want to handle `rocr-ext`? This is very often a key issue for OpenCL development since it is required for Image support. The huge issue is that it is not open source. If we don't have it here, I worry that would-be nixpkgs users of OpenCL would be forced to turn to `nixos-rocm` before they are able to do anything. Additionally, support for `darktable`, for example, requires that image support. A large swath of the broader OpenCL ecosystem needs image support, and we do have it in `nixos-rocm`, but with the caveat that there is the proprietary software question.

We should probably package that. In general, closed-source software is ok for inclusion, but we cannot let ofborg/Hydra build it. We do have to check the EULA to see what the exact requirements are (e.g. some packages like Oracle's JDK require the user to download the package themselves and add it to the store because they need to confirm the EULA explicitly). I can look into that after this PR.

>     * We probably need some documentation associated with configuring NixOS to use ROCm. Things like having to reboot to get OpenCL working (I think this is still the case), are perennial gotchas. Perhaps this belongs in Chapter 16 of the nixpkgs manual.

Agreed. Also about configuring NixOS with OpenCL (and third-party Vulkan) in general. Another possibility would be Part II of the NixOS manual:

https://nixos.org/nixos/manual/index.html#ch-configuration

Thanks again for all the comments and feedback!

**Review by acowley — APPROVED** [review]

**Comment by danieldk** [comment]
@matthewbauer since your are a LLVM codeowner, any objections against merging this?

**Comment by danieldk** [comment]
@Flakebi I think you are also an amdgpu user, since you recently contributed the `amdvlk` derivation? Any chance you could also give this PR a look?

**Comment by Flakebi** [comment]
Thanks for the note @danieldk, I compiled two small OpenCL examples from the [ROCm docs](https://rocmdocs.amd.com/en/latest/Programming_Guides/Opencl-programming-guide.html#example) and it worked great (once I set `OCL_ICD_VENDORS=results/rocm-opencl-icd/etc/OpenCL/vendors/`, the trailing slash is important).

Thank you for your work!

<sup>
> I think you are also an amdgpu user

I guess I’m a little bit more than a user ;)
</sup>

**Comment by danieldk** [comment]
> Thanks for the note @danieldk, I compiled two small OpenCL examples from the [ROCm docs](https://rocmdocs.amd.com/en/latest/Programming_Guides/Opencl-programming-guide.html#example) and it worked great (once I set `OCL_ICD_VENDORS=results/rocm-opencl-icd/etc/OpenCL/vendors/`, the trailing slash is important).

Thanks for testing!

> I guess I’m a little bit more than a user ;)

Whoops, didn't notice that you are involved in GPUOpen. :astonished: :smiley:

**Comment by acowley** [comment]
@Flakebi You should be able to forego that environment variable if you [https://github.com/acowley/nixos-rocm#installation](install the package), specifically the `hardware.opengl.extraPackages = [ pkgs.rocm-opencl-icd ]` part.

**Comment by danieldk** [comment]
Two thumbs up, no complaints, this should be ready to go then.

**Pull request merged**
